Realm/RLMSchema.mm

```diff
@@ -43,172 +43,207 @@ @interface RLMSchema ()
-            RLMReplaceSharedSchemaMethodWithBlock(cls, ^RLMObjectSchema *(Class cls) {
-                if (threadID == pthread_mach_thread_np(pthread_self())) {
-                    return nil;
-                }
-                @synchronized(s_localNameToClass) {
-                    return [cls sharedSchema];
-                }
-            });
+            if (!schema->_objectSchemaByName[prop.objectClassName]) {
+                [errors addObject:[NSString stringWithFormat:@"- '%@.%@' links to class '%@', which is missing from the list of classes to persist", objectSchema.className, prop.name, prop.objectClassName]];
+            }
         }
+    }];
+    if (errors.count) {
+        @throw RLMException(@"Invalid class subset list:\n%@", [errors componentsJoinedByString:@"\n"]);
+    }
 
-        NSMutableArray *schemas = [NSMutableArray arrayWithCapacity:newClasses.count];
-        for (Class cls in newClasses) {
-            RLMObjectSchema *schema = [RLMObjectSchema schemaForObjectClass:cls];
-
-            // set standalone class on shared shema for standalone object creation
-            schema.standaloneClass = RLMStandaloneAccessorClassForObjectClass(schema.objectClass, schema);
-
-            // override sharedSchema classs methods for performance
-            RLMReplaceSharedSchemaMethod(cls, schema);
+    return schema;
+}
 
-            if ([cls shouldIncludeInDefaultSchema]) {
-                [schemas addObject:schema];
-            }
++ (RLMObjectSchema *)sharedSchemaForClass:(Class)cls {
+    @synchronized(s_localNameToClass) {
+        // We create instancse of Swift objects during schema init, and they
+        // obviously need to not also try to initialize the schema
+        if (s_sharedSchemaState == SharedSchemaState::Initializing) {
+            return nil;
         }
 
-        // protected by the @synchronized around s_localNameToClass
-        s_partialSharedSchema.objectSchema = [[s_partialSharedSchema objectSchema] arrayByAddingObjectsFromArray:schemas] ?: schemas;
+        RLMRegisterClassLocalNames(&cls, 1);
+        return RLMRegisterClass(cls);
     }
 }
 
++ (instancetype)partialSharedSchema {
+    return s_sharedSchema;
+}
+
 // schema based on runtime objects
 + (instancetype)sharedSchema {
-    static mach_port_t threadID;
-    if (pthread_mach_thread_np(pthread_self()) == threadID) {
-        @throw RLMException(@"Illegal recursive call of +[%@ %@]. Note: Properties of Swift `Object` classes must not be prepopulated with queried results from a Realm.", self, NSStringFromSelector(_cmd));
-    }
-    static std::once_flag onceFlag;
-    std::call_once(onceFlag, [&](){
-        threadID = pthread_mach_thread_np(pthread_self());
-        RLMSchema *schema = [[RLMSchema alloc] init];
+    @synchronized(s_localNameToClass) {
+        if (s_sharedSchemaState == SharedSchemaState::Initialized) {
+            // Init was done on a different thread while we were waiting for
+            // the @synchronized
```

Comment:
It took me a moment to figure out that this code is indeed specific to when we're waiting on the lock. It isn't apparent at this point in the function that the implementation of `+sharedSchema` is overridden when initialization completes, and therefore this code won't be called if initialization was already complete when `+sharedSchema` was called.